### PR TITLE
Add spec for intersection types

### PR DESCRIPTION
**Pull request by liufengyun**
Add spec for intersection types

### CODE REVIEW

File: docs/docs/reference/intersection-types-spec.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+trait B {
+  def children: List[B]
+}
+val x: A & B = new C
+val ys: List[A & B] = x.children
+```
+
+The type of `children` in `A & B` is the intersection of `children`'s
+type in `A` and its type in `B`, which is `List[A] & List[B]`. This
+can be further simplified to `List[A & B]` because `List` is
```
[the remark this answers is not in the document]
Comment (odersky) [reply docs/docs/reference/intersection-types-spec.md]: I proposed a rule in the subtyping section

File: docs/docs/reference/intersection-types-spec.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+                A <: T
+            ----------------
+               A & B <: T
+
+                B <: T
+            ----------------
+               A & B <: T
+```
+
```
Comment (odersky) [inline docs/docs/reference/intersection-types-spec.md]: If `C` is a type constructor, the join `C[A] & C[B]` is simplified by pulling the intersection inside the constructor, using the following rules:
 
 - If `C` is covariant, `C[A] & C[B] = C[A & B]`
 - If `C` is contravariant, `C[A] & C[B] = C[A | B]`

File: docs/docs/reference/intersection-types-spec.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+```
+Type              ::=  ...| InfixType
+InfixType         ::=  RefinedType {id [nl] RefinedType}
+RefinedType       ::=  WithType {[nl] Refinement}
+WithType          ::=  AnnotType {‘with’ AnnotType}
+```
+
+## Type Checking
+
+The type `S & T` represents values that are of the type `S` and `T` at the same time.
```
Comment (odersky) [inline docs/docs/reference/intersection-types-spec.md]: I would leave this material in the main page.

File: docs/docs/reference/intersection-types-spec.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+values of the type. At the point where a value is _constructed_, one
+must make sure that all inherited members are correctly defined.
+So if one defines a class `C` that inherits `A` and `B`, one needs
+to give at that point a definition of a `children` method with the required type.
+
+```scala
+class C extends A with B {
+  def children: List[A & B] = ???
+}
+```
```
Comment (odersky) [inline docs/docs/reference/intersection-types-spec.md]: Integrate the whole type checking section in the main page.

File: docs/docs/reference/intersection-types-spec.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (liufengyun) [reply docs/docs/reference/intersection-types-spec.md]: It seems `List[A & B] <: List[A] & List[B]` can be derived by the normal subtyping rule for type constructors, no more rules required.

```
    A <: A                        B <: B
  ----------                     ---------
  A & B <: A                     A & B < B
----------------------     ----------------------
List[A & B] <: List[A]     List[A & B] <: List[B]
--------------------------------------------------
      List[A & B] <: List[A] & List[B]
```

File: docs/docs/reference/intersection-types-spec.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+In another word, `A & B` is the same type as `B & A`, in the sense that the two types
+have the same values and are subtypes of each other.
+
+If `C` is a type constructor, the join `C[A] & C[B]` is simplified by pulling the
+intersection inside the constructor, using the following two rules:
+
+- If `C` is covariant, `C[A] & C[B] ~> C[A & B]`
+- If `C` is contravariant, `C[A] & C[B] ~> C[A | B]`
+
```
Comment (milessabin) [inline docs/docs/reference/intersection-types-spec.md]: Where does the "simplified" terminology and the `~>` come from? Intuitively it makes sense, but seems redundant with the typing rules stating the same thing immediately below.

File: docs/docs/reference/intersection-types-spec.md (hunk 2 above)
Comment (milessabin) [reply docs/docs/reference/intersection-types-spec.md]: Ahh ... I think my comment above should have been a follow up to this ...

Comment (liufengyun) [reply docs/docs/reference/intersection-types-spec.md]: Good catch. The derivation provides partial justification for the simplification, as we cannot derive in the other direction.

I tried to find examples where the rules are unsound, but failed. It seems the other direction can be justified informally. @odersky mentioned some theoretical work is needed here based on DOT.

**Pull request merged**